### PR TITLE
Device Defender Tests

### PR DESCRIPTION
**Pull request by shangabl**
### Motivation
Needed to make Device Defender more testable and allow mocking for further development.

#### Change summary
Refactored DD for testability. Added additional unit tests.

### Testing
unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/devicedefender/ReportTaskWrapper.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Aws
+{
+    namespace Iot
+    {
+        namespace DeviceClient
+        {
+            namespace DeviceDefender
+            {
+
```
Comment (marcoemorais-aws) [inline source/devicedefender/ReportTaskWrapper.h:20]: Comments please.

File: test/jobs/TestJobsFeature.cpp [hunk 2]
```diff
@@ -259,7 +259,6 @@ class TestJobsFeature : public ::testing::Test
     void SetUp()
     {
         ThingName = Aws::Crt::String("thing-name value");
-        notifier = shared_ptr<MockNotifier>(new MockNotifier());
```
Comment (marcoemorais-aws) [inline test/jobs/TestJobsFeature.cpp:262]: How is this change related to DD?

**Review by marcoemorais-aws — CHANGES_REQUESTED** [review]

File: test/jobs/TestJobsFeature.cpp (hunk 2 above)
Comment (shangabl) [reply test/jobs/TestJobsFeature.cpp:262]: I must have accidentally deleted a line while referencing Jobs tests.

File: source/devicedefender/ReportTaskWrapper.h (hunk 1 above)
Comment (shangabl) [reply source/devicedefender/ReportTaskWrapper.h:20]: As in for the doxygen? They're there I think at bottom?

File: source/devicedefender/ReportTaskWrapper.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef AWS_IOT_DEVICE_CLIENT_REPORTTASKWRAPPER_H
+#define AWS_IOT_DEVICE_CLIENT_REPORTTASKWRAPPER_H
+
+#include <aws/iotdevicedefender/DeviceDefender.h>
+
+using namespace std;
```
[the remark this answers is not in the document]
Comment (shangabl) [reply source/devicedefender/ReportTaskWrapper.h]: Read through, thanks!

File: source/devicedefender/ReportTaskWrapper.h (hunk 1 above)
Comment (marcoemorais-aws) [reply source/devicedefender/ReportTaskWrapper.h]: I don't see any comments in source/devicedefender/ReportTaskWrapper.h 
There should be some description at least.

**Review by marcoemorais-aws — APPROVED** [review]

**Review by RogerZhongAWS — APPROVED** [review]

File: test/devicedefender/TestDeviceDefender.cpp [hunk 4]
```diff
@@ -86,6 +108,48 @@ TEST_F(TestDeviceDefender, Init)
...
+{
+    /**
+     * Start and Stop test
+     * Inject Report Task, init, and start, verify Thing Name and Interval
+     * Verify calls on Mocks and Assert Start and Stop return 0
+     */
+    EXPECT_CALL(*task, StartTask()).Times(1).WillOnce(Return(0));
+    EXPECT_CALL(*deviceDefender, createReportTask()).Times(1).WillOnce(Return(task));
+    EXPECT_CALL(*deviceDefender, subscribeToTopicFilter()).Times(1).WillOnce(Return());
+    EXPECT_CALL(*notifier, onEvent(_, _)).Times(2).WillOnce(Return());
```
Comment (BolongZhang-AWS) [inline test/devicedefender/TestDeviceDefender.cpp:146]: Is it possible to also verify the exact status like `FEATURE_STARTED` sent in the notification?
